### PR TITLE
[wip] Pydantic v2

### PR DESCRIPTION
**Pull request by benedikt-bartscher**
#1539   
Heavily inspired by @masenf's pydantic branch (https://github.com/reflex-dev/reflex/tree/masenf/pydantic-v2)

### CODE REVIEW

File: reflex/components/component.py [hunk 1]
```diff
@@ -271,11 +277,18 @@ def __init__(self, *args, **kwargs):
                 field_type = EventChain
             elif key in props:
                 # Set the field type.
-                field_type = fields[key].type_
+                field_type = fields[key].annotation
```
Comment (masenf) [inline reflex/components/component.py:280]: in other places in the code `annotation` seems to be a function? 🤔 not sure if this is incorrect, but leaving myself a bookmark with this comment

File: reflex/components/component.py [hunk 2]
```diff
@@ -286,7 +299,7 @@ def __init__(self, *args, **kwargs):
                     if kwargs[key] is None:
                         raise TypeError
 
-                    expected_type = fields[key].outer_type_.__args__[0]
+                    expected_type = field_type.__args__[0]
```
Comment (masenf) [inline reflex/components/component.py:302]: we should probably change this to `typing.get_args(field_type)[0]` while refactoring

File: reflex/components/component.py [hunk 3]
```diff
@@ -520,7 +533,7 @@ def _render(self, props: dict[str, Any] | None = None) -> Tag:
             The tag to render.
         """
         # Create the base tag.
-        tag = Tag(
+        tag: str = Tag(
```
Comment (masenf) [inline reflex/components/component.py:536]:
```suggestion
        tag: Tag = Tag(
```

probably a regex replace issue sneaking in

File: reflex/components/component.py [hunk 4]
```diff
@@ -700,7 +713,7 @@ def render(self) -> Dict:
         Returns:
             The dictionary for template of component.
         """
-        tag = self._render()
+        tag: str = self._render()
```
Comment (masenf) [inline reflex/components/component.py:716]:
```suggestion
        tag: Tag = self._render()
```

File: reflex/components/component.py [hunk 5]
```diff
@@ -1262,7 +1275,7 @@ def __init__(self, *args, **kwargs):
         self.style = Style()
 
         # Set the tag to the name of the function.
-        self.tag = format.to_title_case(self.component_fn.__name__)
+        self.tag: str = format.to_title_case(self.component_fn.__name__)
```
Comment (masenf) [inline reflex/components/component.py:1278]:
```suggestion
        self.tag = format.to_title_case(self.component_fn.__name__)
```

File: reflex/components/core/cond.py [hunk 6]
```diff
@@ -81,7 +81,7 @@ def render(self) -> Dict:
         Returns:
             The dictionary for template of component.
         """
-        tag = self._render()
+        tag: str = self._render()
```
Comment (masenf) [inline reflex/components/core/cond.py:84]:
```suggestion
        tag = self._render()
```

File: reflex/components/core/foreach.py [hunk 7]
```diff
@@ -94,7 +94,7 @@ def render(self):
         Returns:
             The dictionary for template of component.
         """
-        tag = self._render()
+        tag: str = self._render()
```
Comment (masenf) [inline reflex/components/core/foreach.py:97]:
```suggestion
        tag = self._render()
```

File: reflex/components/core/match.py [hunk 8]
```diff
@@ -264,7 +263,7 @@ def render(self) -> Dict:
         Returns:
             The dictionary for template of component.
         """
-        tag = self._render()
+        tag: str = self._render()
```
Comment (masenf) [inline reflex/components/core/match.py:266]:
```suggestion
        tag = self._render()
```

File: reflex/components/markdown/markdown.py [hunk 9]
```diff
@@ -300,7 +300,7 @@ def _get_custom_code(self) -> str | None:
         """
 
     def _render(self) -> Tag:
-        tag = (
+        tag: str = (
```
Comment (masenf) [inline reflex/components/markdown/markdown.py:303]:
```suggestion
        tag = (
```

File: reflex/components/radix/primitives/form.py [hunk 10]
```diff
@@ -116,7 +115,7 @@ def _get_hooks(self) -> str | None:
         )
 
     def _render(self) -> Tag:
-        render_tag = (
+        render_tag: str = (
```
Comment (masenf) [inline reflex/components/radix/primitives/form.py:118]:
```suggestion
        render_tag = (
```

File: reflex/components/radix/themes/base.py [hunk 11]
```diff
@@ -186,7 +189,7 @@ def _get_imports(self) -> imports.ImportDict:
         )
 
     def _render(self, props: dict[str, Any] | None = None) -> Tag:
-        tag = super()._render(props)
+        tag: str = super()._render(props)
```
Comment (masenf) [inline reflex/components/radix/themes/base.py:192]:
```suggestion
        tag = super()._render(props)
```

File: reflex/model.py [hunk 12]
```diff
@@ -61,12 +61,11 @@ def __init_subclass__(cls):
...
             field_name
-            for field_name, field in cls.__fields__.items()
-            if field_name != "id"
-            and getattr(field.field_info, "primary_key", None) is True
+            for field_name, field in cls.model_fields.items()
+            if field_name != "id" and getattr(field, "primary_key", None) is True
         ]
         if non_default_primary_key_fields:
-            cls.__fields__.pop("id", None)
+            cls.model_fields.pop("id", None)
```
Comment (masenf) [inline reflex/model.py:68]: do we need to rebuild the model here?

File: reflex/state.py [hunk 13]
```diff
@@ -1619,7 +1622,22 @@ def _clean(self):
...
+            value: The value of the field.
+
+        Returns:
+            The value of the field.
+        """
+        if isinstance(value, MutableProxy):
+            return super().get_value(value.__wrapped__)
+        return super().get_value(value)
+
+    def _get_value(self, value: Any, **kwargs) -> Any:
```
Comment (masenf) [inline reflex/state.py:1640]: why a new function? is this just WiP artifacts?

File: tests/components/forms/test_debounce.py [hunk 14]
```diff
@@ -43,7 +43,7 @@ def on_change(self, v: str):
 
 def test_render_child_props():
     """DebounceInput should render props from child component."""
-    tag = rx.debounce_input(
+    tag: str = rx.debounce_input(
```
Comment (masenf) [inline tests/components/forms/test_debounce.py:46]:
```suggestion
    tag = rx.debounce_input(
```

the other similar changes in this file should also be reverted

File: tests/components/test_tag.py [hunk 15]
```diff
@@ -52,7 +52,7 @@ def test_is_valid_prop(prop: Var, valid: bool):
 
 def test_add_props():
     """Test that the props are added."""
-    tag = Tag().add_props(key="value", key2=42, invalid=None, invalid2={})
+    tag: str = Tag().add_props(key="value", key2=42, invalid=None, invalid2={})
```
Comment (masenf) [inline tests/components/test_tag.py:55]:
```suggestion
    tag= Tag().add_props(key="value", key2=42, invalid=None, invalid2={})
```

similarly, revert the other similar changes in this file

**Review by masenf — COMMENTED** [review]
Review checkpoint 1: Complete ✅

the `"model_post_init"` in the State was one thing that tripped me up last time i was working in this code.

File: reflex/components/component.py (hunk 1 above)
Comment (benedikt-bartscher) [reply reflex/components/component.py:280]: Yeah, `cs_obj = field.annotation()` is there to initialize default values iirc. Maybe pydantic provides smth like `get_field_default` which handles `default` and `default_factorys` for us?

File: reflex/state.py (hunk 13 above)
Comment (benedikt-bartscher) [reply reflex/state.py:1640]: `_get_value` is there to allow unwrapping the `MutableProxy`s in pydantics serialization. There should be a better way to do this.
Funny sidenote: While `_get_value` is deprecated by pydantic, they somehow seem to rely on it for serialization.

File: reflex/model.py (hunk 12 above)
Comment (benedikt-bartscher) [reply reflex/model.py:68]: Good catch, I think we always need to rebuild when modifying `model_fields`

**Comment by benedikt-bartscher** [comment]
@masenf should we revisit this? Or do you plan to migrate `rx.Base` to dataclasses which could make this obsolete.

**Comment by benedikt-bartscher** [comment]
closing since there seems to be no activity/interest in this
would still love to see reflex using dataclasses or pydantic v2 instead of pydantic v1 for rx.Base

**Pull request closed without merging**